lib/src/component_declaration/component_base_2.dart

```diff
@@ -58,42 +108,101 @@ part of over_react.component_declaration.component_base;
 ///
 /// > Related: [UiStatefulComponent2]
 abstract class UiComponent2<TProps extends UiProps> extends react.Component2
-    with _DisposableManagerProxy
+    with DisposableManagerProxy, GeneratedClass
     implements UiComponent<TProps> {
+  // ***************************************************************************
+  //
+  // Typed props
+  //
+  // ***************************************************************************
+
+  /// A typed view into the component's current JS props object.
+  ///
+  /// Created using [typedPropsFactoryJs] and updated whenever props changes.
   @override
-  Map getDefaultProps() => new JsBackedMap();
+  @toBeGenerated
+  TProps get props {
+    // This needs to be a concrete implementation in Dart 2 for soundness;
+    // without it, you get a confusing error. See: https://github.com/dart-lang/sdk/issues/36191
+    throw new UngeneratedError(member: #props,
+        message: GeneratedErrorMessages.component1AnnotationOnComponent2);
+  }
 
-  /// The props for the non-forwarding props defined in this component.
+  /// Returns a typed props object backed by the specified [propsMap].
+  ///
+  /// Required to properly instantiate the generic [TProps] class.
   @override
-  Iterable<ConsumedProps> get consumedProps => null;
+  @toBeGenerated
+  TProps typedPropsFactory(Map propsMap)=> throw new UngeneratedError(member: #typedPropsFactory);
 
-  /// Returns a copy of this component's props with keys found in [consumedProps] omitted.
+  /// Returns a typed props object backed by the specified [propsMap].
   ///
-  /// > Should be used alongside [forwardingClassNameBuilder].
+  /// Required to properly instantiate the generic [TProps] class.
   ///
-  /// > Related [copyUnconsumedDomProps]
+  /// This should be used where possible over [typedPropsFactory] to allow for
+  /// more efficient dart2js output.
+  @toBeGenerated
+  TProps typedPropsFactoryJs(JsBackedMap propsMap) => throw new UngeneratedError(member: #typedPropsFactoryJs,
+      message: GeneratedErrorMessages.component1AnnotationOnComponent2);
+
+  /// Returns a typed props object backed by a new Map.
   ///
-  /// __Deprecated.__ Use [addUnconsumedProps] within
-  /// [modifyProps] (rather than [addProps]) instead.
+  /// Convenient for use with [getDefaultProps].
+  @override
+  TProps newProps() => typedPropsFactoryJs(new JsBackedMap());
+
+  /// Returns the underlying map object of either [UiProps] or [UiState].
   ///
-  /// Replace
+  /// Used to take a typed factory object and get the underlying backing map.
+  Map getBackingMap(Map map) {
```

Comment:
I feel like we should make this a top-level function utility as opposed to adding to Component2's public API, since it doesn't need access to any instance members.

Thoughts?

Reply:
That sounds good to me. I was trying to decide where the best place was an was pretty on the fence, so that makes a lot of sense!